components/style/properties/longhand/inherited_text.mako.rs

```diff
@@ -737,7 +737,194 @@ ${helpers.single_keyword("text-align-last",
+                ShapeKeyword::Sesame =>  if fill { "\u{fe45}" } else { "\u{fe46}" },
+            }
+        }
+    }
+
+    #[inline]
+    pub fn get_initial_value() -> computed_value::T {
+        computed_value::T::None
+    }
+
+    impl ToComputedValue for SpecifiedValue {
+        type ComputedValue = computed_value::T;
+
+        #[inline]
+        fn to_computed_value(&self, context: &Context) -> computed_value::T {
+            match *self {
+                SpecifiedValue::Keyword(ref keyword) => {
+                    let default_shape = if context.style().get_inheritedbox()
+                                                  .clone_writing_mode() == writing_mode::horizontal_tb {
+                        ShapeKeyword::Circle
+                    } else {
+                        ShapeKeyword::Sesame
+                    };
+                    computed_value::T::Keyword(computed_value::Keyword {
+                        fill: keyword.fill.unwrap_or(true),
+                        shape: keyword.shape.unwrap_or(default_shape)
+                    })
+                },
+                SpecifiedValue::None => computed_value::T::None,
+                SpecifiedValue::String(ref s) => {
+                    let string = if s.len() > 1 {
+                        s.chars().nth(0).unwrap().to_string()
+                    } else { s.clone() };
+                    computed_value::T::String(string)
+                }
+            }
+        }
+        #[inline]
+        fn from_computed_value(computed: &computed_value::T) -> Self {
+            match *computed {
+                computed_value::T::Keyword(ref keyword) => SpecifiedValue::Keyword(Keyword {
+                    fill: Some(keyword.fill),
+                    shape: Some(keyword.shape)
+                }),
+                computed_value::T::None => SpecifiedValue::None,
+                computed_value::T::String(ref string) => SpecifiedValue::String(string.clone())
+            }
+        }
+    }
+
+    pub fn parse(_context: &ParserContext, input: &mut Parser) -> Result<SpecifiedValue, ()> {
+        if input.try(|input| input.expect_ident_matching("none")).is_ok() {
+            return Ok(SpecifiedValue::None);
+        }
+
+        if let Ok(s) = input.try(|input| input.expect_string()) {
+            // Handle <string>
+            return Ok(SpecifiedValue::String(s.into_owned()));
+        }
+
+        // Handle a pair of keywords
+        let mut shape = input.try(ShapeKeyword::parse);
+        let fill = if input.try(|input| input.expect_ident_matching("filled")).is_ok() {
+            Some(true)
+        } else if input.try(|input| input.expect_ident_matching("open")).is_ok() {
+            Some(false)
+        } else { None };
+        if let Err(_) = shape {
+            shape = input.try(ShapeKeyword::parse);
+        }
+
+        // At least one of shape or fill must be handled
+        if let (None, Err(_)) = (fill, shape) {
```

Comment:
It shouldn't parse foo since we are just trying to parse shape keywords like this `shape = input.try(ShapeKeyword::parse);`. ShapeKeyword enum is just 5 keywords (dot, circle, double-circle, triangle, sesame)